fixes #41592, #34559

nesanton opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
When both ```name``` and ```xml``` are specified, there might be a mismatch of domain names in
```name``` and inside ```xml```. This code ignores name and prints a warning.

Allows ```autostart``` to be used with ```command``` and ```state```. Allows standalone ```autostart``` tasks.

Fixes: #41592 
Fixes: #34559

Added handling of errors thrown by libvirt when ```command=define```

Added handling of a situation when a repeated ```define``` task overrides the domain.
From libvirt [docs](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainDefineXML):
-- A previous definition for this domain would be overridden if it already exists.
In real world testing with libvirt versions 1.2.17-13, 2.0.0-10 and 3.9.0-14
on qemu and lxc domains results in:
```operation failed: domain '/name/' already exists with /uuid/```
In case a domain would be indeed overwritten, we should try protecting idempotency.
This code should handle such situation, but it needs testing, because for qemu and lxc domains libvirt refuses to re-define. When re-defining fails, task will end with changed=false
Maybe it's worth to fail when running ```define``` on existing domain.

Updated documentation and added some examples.
<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
virt

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

nesanton committed:
autostart cannot be used with command=define

autostart cannot be used with command=define. If it's used with state, it'll be silently ignored.
https://github.com/ansible/ansible/issues/41592

nesanton committed:
guest name should not be used with xml

When both ```name``` and ```xml``` are specified, there might be a mismatch of domain names in
```name``` and inside ```xml```.

Improved error handling and param deps for autostart-only tasks.
Standalone autostart tasks depend on ```name``` and need the domain to be present.

Added handling of errors thrown by libvirt

Handle situation when a repeated ```define``` overrides the domain.
From libvirt [docs](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainDefineXML):
-- A previous definition for this domain would be overridden if it already exists.
In real world testing with libvirt versions 1.2.17-13, 2.0.0-10 and 3.9.0-14
on qemu and lxc domains results in:
operation failed: domain '<name>' already exists with <uuid>
In case a domain would be indeed overwritten, we should protect idempotency.
Maybe it's worth reverting to just not supporting running ```define``` on existing domain.

nesanton committed:
Updates to documentstion and examples

Removed required flag from name. Added description on some option combinations.
Added a few examples.

nesanton committed:
lint issues

nesanton committed:
yaml lint issues in examples

nesanton committed:
lint pep8 issues

Akasurde commented on lib/ansible/modules/cloud/misc/virt.py:
```diff
@@ -23,14 +23,15 @@
-        defined with xml.
-    required: true
+      - name of the guest VM being managed. Almost always required (see examples). When defining a new VM,
+        name is taken from xml.
```
In previous versions, name is a required parameter. So add a note here stating- 

```
C(name) is an optional parameter from version 2.7 and onwards.
```

nesanton replied:
Sure!
It actually was not really required. You could run things like command: list_vms without name.

Akasurde commented on lib/ansible/modules/cloud/misc/virt.py:
```diff
@@ -474,25 +509,55 @@ def core(module):
+                    # there might be a mismatch between quest 'name' in the module and in the xml
+                    module.warn("'xml' is given - ignoring 'name'")
+                # Is there a better way to get 'name' ?
+                import re
```
import module in head section of module (just below `libvirt` import).

nesanton committed:
import facelift

nesanton committed:
docs: note that name is optional

nesanton commented:
@Akasurde, is there anything I should do further with this PR?

ruzickap committed:
Removing required from documentaion of name parameter (#43530)

* Removing required from documentaion of name parameter

* extra text to make clearer when name is required

horseinthesky commented:
Guys merge it pls.

Akasurde commented:
@nesanton There is conflict in one of the file, could you please resolve the conflict ? Thanks.

nesanton committed:
Option name is not required with option xml.

When defining a domain with xml, its name is taaken directly from the xml definition.

nesanton committed:
Revert "Option name is not required with option xml."

This reverts commit 4ac14a622b38438d1385586683a9275e5905254a.

nesanton commented:
Cherry picked the conflicting commit. Doc on option ```name``` there is not exactly correct though. Mine had that issue fixed actually.
Let's merge this before it gets too old

Akasurde commented:
rebuild_merge
